signer/src/fees.rs

```diff
@@ -0,0 +1,355 @@
+use std::future::Future;
+use std::time::Duration;
+
+use serde::Deserialize;
+
+use crate::error::Error;
+
+const FIVE_MINUTES_SECONDS: i64 = 300;
+const REQUEST_TIMEOUT: Duration = Duration::from_secs(10);
+
+/// A struct representing requests to https://bitcoiner.live
+///
+/// The docs for this API can be found at https://bitcoiner.live/doc/api
+#[derive(Debug, Clone)]
+struct BitcoinerLive {
+    base_url: String,
+}
+
+#[derive(Debug, Deserialize)]
+struct BitcoinerLiveResponse {
+    /// Unix timestamp of when the data was last refreshed
+    timestamp: i64,
+    /// The actual fee rate estimates.
+    estimates: BitcoinerLiveEstimates,
+}
+
+/// BitcoinLive gives fee estimates given the target confirmation time in
+/// minutes. This struct represents part of their JSON response when
+/// requesting fee estimates.
+///
+/// In the actual response, there are also estimates for 60, 120, 180 and
+/// 360 minutes.
+#[allow(dead_code)]
+#[derive(Debug, Deserialize)]
+struct BitcoinerLiveEstimates {
+    #[serde(alias = "30")]
+    thirty: BitcoinerLiveFeeEstimate,
+    #[serde(alias = "60")]
+    sixty: BitcoinerLiveFeeEstimate,
+}
+
+#[derive(Debug, Deserialize)]
+struct BitcoinerLiveFeeEstimate {
+    /// estimated fee rate in satoshis per virtual-byte
+    sat_per_vbyte: f64,
+}
+
+/// A struct representing requests to https://mempool.space
+///
+/// The docs for this API can be found at https://mempool.space/docs/api,
+/// while the specific docs for getting recommended fees can be found at
+/// https://mempool.space/docs/api/rest#get-recommended-fees
+#[derive(Debug, Clone)]
+struct MempoolSpace {
+    base_url: String,
+}
+
+#[allow(dead_code)]
+#[derive(Debug, Deserialize)]
+#[serde(rename_all = "camelCase")]
+struct MempoolSpaceResponse {
+    fastest_fee: u64,
+    half_hour_fee: u64,
+    hour_fee: u64,
+    economy_fee: u64,
+    minimum_fee: u64,
+}
+
+/// A struct representing the recommended fee, in sats per vbyte, from a
+/// particular source.
+#[derive(Debug)]
+pub struct FeeEstimate {
+    pub sats_per_vbyte: f64,
+}
+
+pub trait FeeEstimator {
+    fn estimate_fee_rate(
+        &self,
+        client: &reqwest::Client,
+    ) -> impl Future<Output = Result<FeeEstimate, Error>> + Send;
+}
```

Comment:
I don't think a fee estimator necessarily should have to use a client. I would rather see that as a field on the actual implementations. E.g. `struct MempoolSpaceFeeEstimator { base_url: ..., client: ... }`. This would allow for more flexibility in code depending on the trait. It would be easy to just implement a `ConstantFeeEstimator` for tests that doesn't need to do more fancy fee estimation.

Therefore I think the trait should look like this:

```suggestion
pub trait FeeEstimator {
    fn estimate_fee_rate(
        &self,
    ) -> impl Future<Output = Result<FeeEstimate, Error>> + Send;
}
```

(The receiver could also be `self` and not `&self` which would give the implementer even more flexibility, but this can also lead to some pain with trait bounds. It's not obvious to me what's best here right now)

Reply:
Also, if you don't mind I'd like to follow the trait-as-verb convention and rename this to `EstimateFee`.
 
```suggestion
pub trait EstimateFee {
    fn estimate_fee_rate(
        &self,
        client: &reqwest::Client,
    ) -> impl Future<Output = Result<FeeEstimate, Error>> + Send;
}
```